File: eval/benchmarks/jacobi-2d-imper/Jacobi2DImper.java
public class Jacobi2DImper {
    static final int N = 64;
    static final int TSTEPS = 20;
    static long[][] a = new long[N][N];
    static long[][] b = new long[N][N];

    static void initArray() {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                a[i][j] = (i * (j + 2) + 2) / N;
                b[i][j] = (i * (j + 3) + 3) / N;
            }
        }
    }

    static void kernel() {
        for (int t = 0; t < TSTEPS; t++) {
            for (int i = 1; i < N - 1; i++) {
                for (int j = 1; j < N - 1; j++) {
                    b[i][j] = (a[i][j] + a[i][j - 1] + a[i][j + 1] + a[i + 1][j] + a[i - 1][j]) / 5;
                }
            }
            for (int i = 1; i < N - 1; i++) {
                for (int j = 1; j < N - 1; j++) {
                    a[i][j] = b[i][j];
                }
            }
        }
    }

    static long checksum() {
        long s = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                s += a[i][j];
            }
        }
        return s;
    }

    public static void main(String[] args) {
        initArray();
        kernel();
        System.err.println("checksum=" + checksum());
    }
}

File: eval/benchmarks/jacobi-2d-imper/jacobi_2d_imper.rs
const N: usize = 64;
const TSTEPS: usize = 20;

fn init_array(a: &mut [[i64; N]; N], b: &mut [[i64; N]; N]) {
    for i in 0..N {
        for j in 0..N {
            a[i][j] = ((i * (j + 2)) as i64 + 2) / N as i64;
            b[i][j] = ((i * (j + 3)) as i64 + 3) / N as i64;
        }
    }
}

fn kernel(a: &mut [[i64; N]; N], b: &mut [[i64; N]; N]) {
    for _ in 0..TSTEPS {
        for i in 1..N - 1 {
            for j in 1..N - 1 {
                b[i][j] = (a[i][j] + a[i][j - 1] + a[i][j + 1] + a[i + 1][j] + a[i - 1][j]) / 5;
            }
        }
        for i in 1..N - 1 {
            for j in 1..N - 1 {
                a[i][j] = b[i][j];
            }
        }
    }
}

fn checksum(a: &[[i64; N]; N]) -> i64 {
    let mut s = 0;
    for row in a {
        for v in row {
            s += *v;
        }
    }
    s
}

fn main() {
    let mut a = [[0i64; N]; N];
    let mut b = [[0i64; N]; N];
    init_array(&mut a, &mut b);
    kernel(&mut a, &mut b);
    eprintln!("checksum={}", checksum(&a));
}
